cli/node/analyze.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+use std::collections::HashSet;
+
+use deno_ast::swc::common::SyntaxContext;
+use deno_ast::view::Node;
+use deno_ast::view::NodeTrait;
+use deno_ast::ModuleSpecifier;
+use deno_ast::ParsedSource;
+use deno_ast::SourceRanged;
+use deno_core::error::AnyError;
+use std::fmt::Write;
+
+static NODE_GLOBALS: &[&str] = &[
+  "Buffer",
+  "clearImmediate",
+  "clearInterval",
+  "clearTimeout",
+  "global",
+  "process",
+  "setImmediate",
+  "setInterval",
+  "setTimeout",
+];
+
+// todo(dsherret): this code is way more innefficient than it needs to be.
+//
+// In the future, we should disable capturing tokens & scope analysis
+// and instead only use swc's APIs to go through the portions of the tree
+// that we know will affect the global scope while still ensuring that
+// `var` decls are taken into consideration.
+
+pub fn esm_code_with_node_globals(
+  specifier: &ModuleSpecifier,
+  code: String,
```

Comment:
I think `&str` would be better here

Reply:
The `deno_ast::SourceTextInfo::from_string(code)` below takes ownership of the string.